oci/layer/generate.go

```diff
@@ -97,3 +100,44 @@ func GenerateLayer(path string, deltas []mtree.InodeDelta, opt *MapOptions) (io.
 
 	return reader, nil
 }
+
+// GenerateInsertLayer generates a completely new layer from the root path to
+// be inserted into the image at "prefix".
+func GenerateInsertLayer(root string, prefix string, opt *MapOptions) io.ReadCloser {
+	root = path.Clean(root)
+	rootPrefixLen := len(root) - len(path.Base(root))
+	prefix = strings.TrimPrefix(prefix, "/")
+
+	var mapOptions MapOptions
+	if opt != nil {
+		mapOptions = *opt
+	}
+
+	reader, writer := io.Pipe()
+
+	go func() {
+		var err error
+
+		defer func() {
+			writer.CloseWithError(errors.Wrap(err, "generate layer"))
+		}()
+
+		tg := newTarGenerator(writer, mapOptions)
+
+		err = filepath.Walk(root, func(curPath string, info os.FileInfo, err error) error {
```

Comment:
I think this should be using `unpriv.Walk` (with the regular checks for which `fs` we should use). One of the things people want is for `umoci` to use it's magic `CAP_DAC_OVERRIDE` emulation to be able to handle annoying paths.

And once we add that, we'll need another integration test to make sure that the path modes are included correctly.